File: BOJ/27_동적 계획법과 최단거리 역추적/14003_LIS5.cpp
#include <cstdio>
#include <vector>

#define MAXN 1000000

using namespace std;

int nums[MAXN+1];
vector<int> lis,ans,idxs;

int N;

int bisrch(int n) {
    int left = 0, right = lis.size()-1;

    while(left <= right) {
        int mid = (left + right) / 2;

        if(lis[mid] < n) left = mid + 1;
        else right = mid - 1;
    }
    return right + 1;
}

int main() {

    scanf("%d",&N);
    for(int i=0; i<N; i++) scanf("%d",nums+i);

    int cnt = 0;
    for(int i=0; i<N; i++) {

        if(lis.empty() || lis[cnt-1] < nums[i]) {
            lis.push_back(nums[i]);
            cnt++;
            idxs.push_back(lis.size()-1);
        }
        else {
            int idx = bisrch(nums[i]);
            lis[idx] = nums[i];
            idxs.push_back(idx);
        }
    }
    for(int i=idxs.size()-1; i>=0; i--) {
        if(idxs[i] == cnt-1) {
            ans.push_back(nums[i]);
            cnt--;
        }
    }
    printf("%lu\n",ans.size());
    for(int i=ans.size()-1; i>=0; i--) printf("%d ",ans[i]);
    return 0;
}

File: BOJ/27_동적 계획법과 최단거리 역추적/14003_LIS5.py
import sys

read = sys.stdin.readline

#def 풀이과정 :
    # O(NlogN) LIS 구하기
    # 이분 탐색 통해서 lower_bound 구해서 풀어줌.

def bisrch(dp,n) :
    left = 0; right = len(dp)
    while(left <= right) :
        mid = (left+right) // 2
        if dp[mid] < n :
            left = mid + 1
        else :
            right = mid - 1
    return right + 1


N = int(read())
nums = list(map(int,read().split()))
dp = []
idxs = []
for i in range(N) :
    if not dp or dp[-1] < nums[i] :
        dp.append(nums[i])
        idxs.append(len(dp)-1)
    else :
        idx = bisrch(dp, nums[i])
        dp[idx] = nums[i]
        idxs.append(idx)

ans = []
tmp = len(dp)-1
for i in range(N-1,-1,-1):
    if idxs[i] == tmp :
        ans.append(nums[i])
        tmp -= 1

print(len(ans))
print(*reversed(ans))